[Offload] Move `/openmp/libomptarget` to `/offload`

jdoerfert opened the pull request:
In a nutshell, this moves our libomptarget code to create the offload subproject.

With this commit, users need to enable the new LLVM/Offload subproject as a runtime in their cmake configuration.
No further changes are expected for downstream code.

Tests and other components still depend on OpenMP and have also not been renamed. The results below are for a build in which OpenMP and Offload are enabled runtimes. In addition to the pure `git mv`, I needed to adjust some CMake files. Nothing is intended to change semantics.

```
ninja check-offload
```
Works with the X86 and AMDGPU offload tests

```
ninja check-openmp
```
Still works but doesn't build offload tests anymore.

```
ls install/lib
```
Shows all expected libraries, incl.
- `libomptarget.devicertl.a`
- `libomptarget-nvptx-sm_90.bc`
- `libomptarget.rtl.amdgpu.so` -> `libomptarget.rtl.amdgpu.so.18git`
- `libomptarget.so` -> `libomptarget.so.18git`

Fixes: https://github.com/llvm/llvm-project/issues/75124

jdoerfert committed:
[Offload] Move `/openmp/libomptarget` to `/offload`

In a nutshell, this moves our libomptarget code to create the offload
subproject.

With this commit, users need to enable the new LLVM/Offload subproject
as a runtime in their cmake configuration. No further changes are
expected for downstream code.

Tests and other components still depend on OpenMP and have also not been
renamed. The results below are for a build in which OpenMP and Offload
are enabled runtimes. In addition to the pure git mv, I needed to adjust
some CMake files. Nothing is intended to change semantics.

```
ninja check-offload
```
Works with the X86 and AMDGPU offload tests

```
ninja check-openmp
```
Still works but doesn't build offload tests anymore.

```
ls install/lib
```
Shows all expected libraries, incl.
- `libomptarget.devicertl.a`
- `libomptarget-nvptx-sm_90.bc`
- `libomptarget.rtl.amdgpu.so` -> `libomptarget.rtl.amdgpu.so.19git`
- `libomptarget.so` -> `libomptarget.so.19git`

Fixes: https://github.com/llvm/llvm-project/issues/75124

doru1004 commented:
Libomptarget has been written for many years by many people. I would like to understand, does GH allow you to transfer over the history of contributions for each file you are moving?

ldionne commented:
> Libomptarget has been written for many years by many people. I would like to understand, does GH allow you to transfer over the history of contributions for each file you are moving?

Git normally handles file renaming quite well, so I would assume you won't even notice this when doing e.g. `git blame`.

The bigger question I'd have is about creating a new top-level directory in the monorepo -- typically this has been done mainly for new projects (e.g. llvm-libc, mlir, etc) and I wonder if there's some sort of process to add a new one of those. I don't care strongly, just mentioning it.

doru1004 commented:
> Git normally handles file renaming quite well, so I would assume you won't even notice this when doing e.g. `git blame`.

Thank you for clarifying, sounds good.

jplehr commented:
I gave this a shot with our buildbot CMake invocation to see what would happen and what we need to change, once this lands.
The CMake line the bot uses was modified to add the `offload` runtime, all else is the same.
```
cmake -G Ninja ../llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=ON '-DLLVM_LIT_ARGS=-vv --show-unsupported --show-xfail -j 64' -DCMAKE_INSTALL_PREFIX=/tmp/llvm.inst -DCMAKE_BUILD_TYPE=Release '-DLLVM_TARGETS_TO_BUILD=X86;AMDGPU' '-DLLVM_ENABLE_RUNTIMES=offload;openmp' '-DLLVM_ENABLE_PROJECTS=clang;lld;llvm;compiler-rt'
```

I get errors about not finding something
```
CMake Error at $HOME/git/llvm-project/offload/CMakeLists.txt:126 (include):
  include could not find requested file:

    OpenMPTesting
```

I also get the following that I'm not sure what to do about (I don't get this in our regular buildbot case)
```
CMake Warning at $HOME/git/llvm-project/offload/CMakeLists.txt:182 (message):
  LTO is not supported: Change Dir:
  $HOME/git/llvm-project/botbuild/runtimes/runtimes-bins/offload/CMakeFiles/_CMakeLTOTest-CXX/bin

  Run Build Command(s):/usr/bin/ninja && [1/4] Building CXX object
  CMakeFiles/foo.dir/foo.cpp.o

  [2/4] Building CXX object CMakeFiles/boo.dir/main.cpp.o

  [3/4] Linking CXX static library libfoo.a

  [4/4] Linking CXX executable boo

  FAILED: boo

  : && $HOME/git/llvm-project/botbuild/./bin/clang++ -flto=thin
  CMakeFiles/boo.dir/main.cpp.o -o boo libfoo.a && :

  /usr/bin/ld:
  $HOME/git/llvm-project/botbuild/bin/../lib/LLVMgold.so: error
  loading plugin:
  $HOME/git/llvm-project/botbuild/bin/../lib/LLVMgold.so: cannot
  open shared object file: No such file or directory

  clang++: error: linker command failed with exit code 1 (use -v to see
  invocation)

  ninja: build stopped: subcommand failed.
```

jdoerfert commented:
The first one I'll look into.
The second one is something you should see already in trunk. It's the test for LTO, though, arguably, I should not just emit the entire warning if it failed. I'll cut that down.

jplehr commented:
I tested this again with the buildbot config that we use and it seems that OMPT support got dropped.

In current libomptarget we have this logic https://github.com/llvm/llvm-project/blob/fa2bbea14df3273b3403f34cc295c56233fdbd0d/openmp/libomptarget/CMakeLists.txt#L104C1-L120C8 but that is not present in the new CMake file.

The result is that OMPT is not built by default, as it is today. I am not sure if we have a switch to turn it on in offload. In any case, I feel that we should aim to maintain the defaults that we have today, i.e., OMPT support on-by-default.

saiislam commented:
> I tested this again with the buildbot config that we use and it seems that OMPT support got dropped.
> 
> In current libomptarget we have this logic https://github.com/llvm/llvm-project/blob/fa2bbea14df3273b3403f34cc295c56233fdbd0d/openmp/libomptarget/CMakeLists.txt#L104C1-L120C8 but that is not present in the new CMake file.
> 
> The result is that OMPT is not built by default, as it is today. I am not sure if we have a switch to turn it on in offload. In any case, I feel that we should aim to maintain the defaults that we have today, i.e., OMPT support on-by-default.

One way to solve it is to propagate the value of `LIBOMP_HAVE_OMPT_SUPPORT` to llvm-project/offload/CMakeLists.txt using:

`set(LIBOMP_HAVE_OMPT_SUPPORT ${LIBOMP_HAVE_OMPT_SUPPORT} PARENT_SCOPE)`

Above line needs to be added to [openmp/CMakeLists.txt](https://github.com/llvm/llvm-project/blob/5ad320abe36357e3290007d3ab353e8637f33720/openmp/CMakeLists.txt#L125) and [openmp/runtime/CMakeLists.txt](https://github.com/llvm/llvm-project/blob/5ad320abe36357e3290007d3ab353e8637f33720/openmp/runtime/CMakeLists.txt#L353)

dhruvachak commented:
openmp/runtime/cmake/config-ix.cmake already has the parent_scope setting. 

set(LIBOMP_HAVE_OMPT_SUPPORT ${LIBOMP_HAVE_OMPT_SUPPORT} PARENT_SCOPE)

So the openmp/runtime/CMakeLists.txt should not need a change. But the openmp CMakeLists.txt may need this setting as @saiislam said.

jdoerfert commented:
Added `set(LIBOMP_HAVE_OMPT_SUPPORT ${LIBOMP_HAVE_OMPT_SUPPORT} PARENT_SCOPE)` to the CMake as suggested, also rebased it on trunk.

dhruvachak commented:
@jdoerfert I think you additionally need the cmake changes @jplehr posted a couple of days back. This one:

In current libomptarget we have this logic https://github.com/llvm/llvm-project/blob/fa2bbea14df3273b3403f34cc295c56233fdbd0d/openmp/libomptarget/CMakeLists.txt#L104C1-L120C8
but that is not present in the new CMake file.

Without these changes, I think you will see the OMPT tests as unsupported.

jplehr commented:
> Added `set(LIBOMP_HAVE_OMPT_SUPPORT ${LIBOMP_HAVE_OMPT_SUPPORT} PARENT_SCOPE)` to the CMake as suggested, also rebased it on trunk.

I just did another local test, and the OMPT tests are still shown as `UNSUPPORTED`. I think @dhruvachak is right in that we need that additional logic to actually enable OMPT support in `offload`.

jdoerfert commented:
> @jdoerfert I think you additionally need the cmake changes @jplehr posted a couple of days back. This one:
> 
> In current libomptarget we have this logic https://github.com/llvm/llvm-project/blob/fa2bbea14df3273b3403f34cc295c56233fdbd0d/openmp/libomptarget/CMakeLists.txt#L104C1-L120C8 but that is not present in the new CMake file.
> 
> Without these changes, I think you will see the OMPT tests as unsupported.

That logic was present the entire time in offload/CMakeLists.txt. 
I'll look into it.

dhruvachak commented:
Thanks @jdoerfert 
Yes, I see that change now in offload/CMakeLists.txt. @jplehr Please verify you see the OMPT tests passing now in the buildbot.

saiislam committed:
[Offload] Support standalone build for OpenMP and Offload

The standalone offload build will just search for
the ompt header and if found then OMPT support will
be enabled. This caused the LIBOMP_HAVE_VERSION_SCRIPT_FLAG
detection and pythonize_bool to move into offload/CMakeLists.txt.

If the cmake compiler is clang, get the resource directory for
installation of openmp headers. find_package(LLVM) is already
called in LibomptargetGetDependencies.

Depends on #75125

saiislam reviewed (approved):
LGTM.

Thanks for the support and patience!

jplehr reviewed (approved):
I think we are good from the build bots point of view.

ye-luo commented:
libomptarget files install location is changed back to `<install>/lib` from `<install>/lib/x86_64-unknown-linux-gnu`

jhuber6 commented:
> libomptarget files install location is changed back to `<install>/lib` from `<install>/lib/x86_64-unknown-linux-gnu`

I'm waiting for someone to review https://github.com/llvm/llvm-project/pull/89645

daltenty commented on openmp/CMakeLists.txt:
```diff
@@ -113,7 +113,17 @@ option(OPENMP_ENABLE_LIBOMP_PROFILING "Enable time profiling for libomp." OFF)
+      RESULT_VARIABLE COMMAND_RETURN_CODE
+      OUTPUT_VARIABLE COMPILER_RESOURCE_DIR
+    )
+    set(LIBOMP_HEADERS_INSTALL_PATH "${COMPILER_RESOURCE_DIR}/include")
```
The addition of this change seems to have broken the stand alone build we were doing:
```
-- Installing: /nfs/build//lib/clang/19/include/omp.h
CMake Error at runtime/src/cmake_install.cmake:101 (file):
  file INSTALL cannot copy file
  "/home/daltenty/build-openmp/32/runtime/src/omp.h" to
  "/nfs/build/lib/clang/19/include/omp.h":
  Read-only file system.
Call Stack (most recent call first):
  runtime/cmake_install.cmake:42 (include)
  cmake_install.cmake:42 (include)
  
 We're now trying to install to the build compiler's resource directory for some reason, which is read-only and isn't really what we intended here.

jhuber6 replied:
Yeah that's definitely wrong for a standalone build. For a runtimes build it's so we install it in a place where the compiler can find it so `./bin/clang` actually works. We'll need to fix a lot of this.

daltenty replied:
I think there's some confusion here about the effects of this.

Just for terminology clarity, my understanding is there are roughly three ways we can build runtimes subprojects of LLVM (defined by the top-level CMake source directory and CMake options you use):

1. **Projects+Runtimes build** with Clang (i.e. ):
`cmake -S llvm-project/llvm -DENABLE_PROJECTS=clang -DENABLE_RUNTIMES=openmp`
2. **Runtimes-only build**
`cmake -S llvm-project/runtimes -DENABLE_RUNTIMES=openmp`
3. **Stand-alone build** (from before the monorepo existed, doesn't really work for all subprojects anymore)
`cmake -S llvm-project/openmp -DOPENMP_LLVM_TOOLS_DIR=/path/to/llvm/build/bin`

> Yeah that's definitely wrong for a standalone build. For a runtimes build it's so we install it in a place where the compiler can find it so `./bin/clang` actually works. We'll need to fix a lot of this.

I assume this code was intended for a Projects+Runtimes build with clang. But it's inside a `if(OPENMP_STANDALONE_BUILD)` block which only seems to be true for standalone builds. 

There are some misleading comments in the source about this here https://github.com/llvm/llvm-project/blob/3e921d3c52542b3284c4bd0ef29a70511d38c79c/openmp/CMakeLists.txt#L3 but I checked my builds and I don't see it being set by llvm/runtimes at all and seems to be false during the tests. So I'm not sure this doing what was intended.

daltenty replied:
Posted https://github.com/llvm/llvm-project/pull/91243 to address this

fhahn commented:
With this change, previous cmake configurations only using `-DLLVM_ENABLE_RUNTIMES="openmp"` silently appear to not configure `openmp` at all; if `offload` is required for `openmp` to be enabled, could we error on  `-DLLVM_ENABLE_RUNTIMES="openmp"`?

It also looks like https://openmp.llvm.org/SupportAndFAQ.html needs updating?

jdoerfert commented:
@jhuber6 @jplehr Are you already looking at the reported issues above?
